dhis-2/dhis-services/dhis-service-core/src/main/java/org/hisp/dhis/user/DefaultUserService.java

```diff
@@ -755,18 +759,42 @@ public void enableTwoFa(User user, String code) {
     approveTwoFactorSecret(user);
   }
 
+  @Override
+  public void registerFailed2FADisableAttempt(String username) {
+    Integer attempts = twoFaDisableFailedAttemptCache.get(username).orElse(0);
+    attempts++;
+    twoFaDisableFailedAttemptCache.put(username, attempts);
+  }
+
+  @Override
+  public void registerSuccess2FADisable(String username) {
+    twoFaDisableFailedAttemptCache.invalidate(username);
+  }
+
+  @Override
+  public boolean twoFaDisableIsLocked(String username) {
+    return twoFaDisableFailedAttemptCache.get(username).orElse(0) >= LOGIN_MAX_FAILED_ATTEMPTS;
+  }
+
   @Transactional
   @Override
   public void disableTwoFa(User user, String code) {
     if (user.getSecret() == null) {
       throw new IllegalStateException("Two factor is not enabled, enable first");
     }
 
+    if (twoFaDisableIsLocked(user.getUsername())) {
+      throw new IllegalStateException("Too many failed attempts, try again later");
```

Comment:
could we be more helpful to the user & advise how long they need to wait?

Reply:
Well, a little bit of the point is to avoid attacks, so that would also help the attackers to script it. Classic security vs. convenience dilemma.

Reply:
By not providing that info it does not prevent attacks, only possibly delays them. We're not really preventing anyone obtaining that info given we're open source.